File: core/parser/type_declaration_listener.cpp
/*
 * Copyright (C) 2016 Riccardo De Benedictis <riccardo.debenedictis@istc.cnr.it>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "type_declaration_listener.h"
#include "parser.h"
#include "../core.h"
#include "type_visitor.h"
#include "typedef_type.h"
#include "enum_type.h"

using namespace oratio;

type_declaration_listener::type_declaration_listener(parser * const p) : _parser(p), _scope(p->_core) { }

type_declaration_listener::~type_declaration_listener() { }

void type_declaration_listener::enterCompilation_unit(oRatioParser::Compilation_unitContext* ctx) {
    _parser->_scopes.insert({ ctx, _parser->_core });
}

void type_declaration_listener::enterTypedef_declaration(oRatioParser::Typedef_declarationContext* ctx) {
    // A new typedef type has been declared..
    typedef_type* td = new typedef_type(_parser->_core, _scope, ctx->name->getText(), type_visitor(_parser).visit(ctx->primitive_type()), ctx->expr());
    if (core * c = dynamic_cast<core*> (_scope)) {
        c->_types.insert({ ctx->name->getText(), td });
    }
    else if (type * t = dynamic_cast<type*> (_scope)) {
        t->_types.insert({ ctx->name->getText(), td });
    }
}

void type_declaration_listener::enterEnum_declaration(oRatioParser::Enum_declarationContext* ctx) {
    // A new enum type has been declared..
    enum_type* et = new enum_type(_parser->_core, _scope, ctx->name->getText());
    _parser->_scopes.insert({ ctx, et });

    // We add the enum values..
    for (const auto& c : ctx->enum_constants()) {
        for (const auto& l : c->StringLiteral()) {
            std::string c_enum = l->getText();
            et->add_enum(_parser->_core->new_string(c_enum));
        }
    }
    if (core * c = dynamic_cast<core*> (_scope)) {
        c->_types.insert({ ctx->name->getText(), et });
    }
    else if (type * t = dynamic_cast<type*> (_scope)) {
        t->_types.insert({ ctx->name->getText(), et });
    }
}

void type_declaration_listener::enterClass_declaration(oRatioParser::Class_declarationContext* ctx) {
    // A new type has been declared..
    type* c_t = new type(_parser->_core, _scope, ctx->name->getText());
    _parser->_scopes.insert({ ctx, c_t });
    if (core * e = dynamic_cast<core*> (_scope)) {
        e->_types.insert({ ctx->name->getText(), c_t });
    }
    else if (type * t = dynamic_cast<type*> (_scope)) {
        t->_types.insert({ ctx->name->getText(), c_t });
    }

    _scope = c_t;
}

void type_declaration_listener::exitClass_declaration(oRatioParser::Class_declarationContext* ctx) {
    _scope = _scope->_scope;
}

void type_declaration_listener::enterClass_type(oRatioParser::Class_typeContext* ctx) {
    _parser->_scopes.insert({ ctx, _scope });
}

File: java-api/src/main/java/it/cnr/istc/oratio/riddle/TypeDeclarationListener.java
/*
 * Copyright (C) 2018 Riccardo De Benedictis <riccardo.debenedictis@istc.cnr.it>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package it.cnr.istc.oratio.riddle;

/**
 *
 * @author Riccardo De Benedictis <riccardo.debenedictis@istc.cnr.it>
 */
class TypeDeclarationListener extends riddleBaseListener {

    private final Core core;
    private final riddleParser parser;
    private Scope scope = null;

    TypeDeclarationListener(final Core core, final riddleParser parser) {
        this.core = core;
        this.parser = parser;
        this.scope = core;
    }

    @Override
    public void enterTypedef_declaration(riddleParser.Typedef_declarationContext ctx) {
        defineType(new Typedef(core, scope, new TypeVisitor(core, parser).visit(ctx.primitive_type()),
                ctx.name.getText()));
    }

    @Override
    public void enterEnum_declaration(riddleParser.Enum_declarationContext ctx) {
        // A new enum type has been defined..
        EnumType c_type = new EnumType(core, scope, ctx.name.getText());
        core.scopes.put(ctx, c_type);
        // We add the enum values..
        ctx.enum_constants().forEach(enum_constant -> enum_constant.StringLiteral()
                .forEach(tn -> c_type.vals.add(tn.getSymbol().getText())));
        defineType(c_type);
    }

    @Override
    public void enterClass_declaration(riddleParser.Class_declarationContext ctx) {
        // A new class has been defined..
        Type c_type = new Type(core, scope, ctx.name.getText());
        core.scopes.put(ctx, c_type);
        defineType(c_type);
        scope = c_type;
    }

    @Override
    public void exitClass_declaration(riddleParser.Class_declarationContext ctx) {
        scope = scope.getScope();
    }

    private void defineType(Type type) {
        if (scope instanceof Core)
            ((Core) scope).types.put(type.name, type);
        else if (scope instanceof Type)
            ((Type) scope).types.put(type.name, type);
    }

    @Override
    public void enterClass_type(riddleParser.Class_typeContext ctx) {
        core.scopes.put(ctx, scope);
    }
}
